prow/git/v2/remote.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package git
+
+import (
+	"fmt"
+	"net/url"
+	"path"
+)
+
+// RemoteResolverFactory knows how to construct remote resolvers for
+// authoritative central remotes (to pull from) and publish remotes
+// (to push to) for a repository. These resolvers
+type RemoteResolverFactory interface {
+	// CentralRemote returns a resolver for a remote server with an
+	// authoritative version of the repository. This type of remote
+	// is useful for fetching refs and cloning.
+	CentralRemote(org, repo string) RemoteResolver
+	// PublishRemote returns a resolver for a remote server with a
+	// personal fork of the repository. This type of remote is most
+	// useful for publishing local changes.
+	PublishRemote(org, repo string) RemoteResolver
+}
+
+// RemoteResolver knows how to construct a remote URL for git calls
+type RemoteResolver func() (string, error)
+
+type sshRemoteResolverFactory struct {
+	host     string
+	username func() (login string, err error)
+}
+
+// CentralRemote creates a remote resolver that refers to an authoritative remote
+// for the repository.
+func (f *sshRemoteResolverFactory) CentralRemote(org, repo string) RemoteResolver {
+	remote := fmt.Sprintf("git@%s:%s/%s.git", f.host, org, repo)
+	return func() (string, error) {
+		return remote, nil
+	}
+}
+
+// PublishRemote creates a remote resolver that refers to a gitUser's remote
+// for the repository that can be published to.
+func (f *sshRemoteResolverFactory) PublishRemote(_, repo string) RemoteResolver {
+	return func() (string, error) {
+		org, err := f.username()
+		if err != nil {
+			return "", err
+		}
+		return fmt.Sprintf("git@%s:%s/%s.git", f.host, org, repo), nil
+	}
+}
+
+type simpleAuthResolverFactory struct {
+	host     string
+	username func() (login string, err error)
+	token    func() []byte
```

Comment:
Given that this is casted to `string` anyway, maybe we can require this to return a string?

Reply:
This adapts the censoring executor/logger interface